### PR TITLE
Fine tune references in console.

### PR DESCRIPTION
**Pull request by polina-c**
RELEASE_NOTE_EXCEPTION=[behind flag]

https://user-images.githubusercontent.com/12115586/225167586-49b6faab-e3f5-4d78-8d0e-ec5298a612c2.mov

Screenshot 2023-03-14 at 4 57 47 PM

### CODE REVIEW

File: packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart [hunk 1]
```diff
@@ -63,7 +63,9 @@ class ObjectReferences extends GenericInstanceRef {
     required IsolateRef super.isolateRef,
     required super.value,
     required HeapObjectSelection super.heapSelection,
-  });
+  }) {
+    if (refNodeType.isLive) assert(value != null);
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart:67]: Nit: `assert(!refNodeType.isLive || value != null)`

File: packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart [hunk 2]
```diff
@@ -113,4 +125,6 @@ enum RefNodeType {
   final RefDirection? direction;
 
   bool get isRoot => {refRoot, staticRefRoot, liveRefRoot}.contains(this);
+
+  bool get isLive => {liveOutRefs, liveInRefs, liveRefRoot}.contains(this);
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart:129]: Can these sets be `const`?

File: packages/devtools_app/lib/src/shared/diagnostics/references.dart [hunk 3]
```diff
@@ -87,9 +86,9 @@ Future<void> addChildReferences(
             ),
           ),
       ]);
-      break;
+      return;
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/diagnostics/references.dart:89]: Nit: Are these returns necessary? We won't require `break`s in `case` statements starting with 3.0, so it'd probably be cleaner to leave them as is and then remove them when we can.

File: packages/devtools_app/lib/src/shared/diagnostics/references.dart [hunk 4]
```diff
@@ -279,18 +282,26 @@ void _addLiveReferenceToNode(
   HeapObjectSelection heapSelection, {
   String namePrefix = '',
 }) {
-  if (instance is! InstanceRef && instance is! ContextRef) return;
+  if (instance is! ObjRef) throw 'Unexpected type: ${instance.runtimeType}.';
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/diagnostics/references.dart:285]: Don't throw bare strings. Consider throwing `TypeError` or `ArgumentError` instead.

Comment (bkonyi) [reply packages/devtools_app/lib/src/shared/diagnostics/references.dart:285]: Also, if this is always expected to be an `ObjRef`, the `instance` parameter should have type `ObjRef` or `ObjRef?`. We should avoid using `Object` when we have a known expected type as it's always better to have a static analysis failure than a runtime failure.

File: packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart (hunk 1 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart:67]: Current option looks more readable for me.

File: packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart (hunk 2 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart:129]: yes!

File: packages/devtools_app/lib/src/shared/diagnostics/references.dart (hunk 3 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/diagnostics/references.dart:89]: cool!

In this exact case the method early returns sometimes in the middle of the case. So, I want it to be consistent  and make sure all options do not count on any 'after switch' operations in this method.

File: packages/devtools_app/lib/src/shared/diagnostics/references.dart (hunk 4 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/diagnostics/references.dart:285]: I agree. I am not confident about expectations for this method. I just never saw non ObjRef so far. 
So, If I ever get non ObjRef, I (or user) will get exception and will add processing for it.

Comment (polina-c) [reply packages/devtools_app/lib/src/shared/diagnostics/references.dart]: I dream to refactor this code eventually to make contracts clearer.

File: packages/devtools_app/lib/src/shared/diagnostics/references.dart (hunk 3 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/diagnostics/references.dart:89]: Where can I read more about non-required break? Right now absence of break means  that the case is combined with next one. So, I am wondering how backward compatibility is handled.

File: packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart [hunk 5]
```diff
@@ -83,6 +85,16 @@ class ObjectReferences extends GenericInstanceRef {
...
   IsolateRef get isolateRef => super.isolateRef!;
+
+  int? get childCount {
+    final result = heapSelection.countOfReferences(refNodeType.direction);
+    if (result != null) return result;
+
+    final inatance = value;
+    if (inatance is! InstanceRef) return null;
+
+    return inatance.length;
```
Comment (CoderDake) [inline packages/devtools_app/lib/src/shared/diagnostics/generic_instance_reference.dart:96]:
```suggestion
    final instance = value;
    if (instance is! InstanceRef) return null;

    return instance.length;
```

**Review by kenzieschmoll — APPROVED** [review]

**Pull request merged**